### PR TITLE
NIFI-4256 - Add support for all AWS S3 Encryption Options

### PR DESCRIPTION
**Pull request by baank**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/s3/AbstractS3Processor.java [hunk 1]
```diff
@@ -125,16 +125,31 @@
                     new AllowableValue("S3SignerType", "Signature v2"))
             .defaultValue("Default Signature")
             .build();
+
+    public static final PropertyDescriptor CLIENT_SERVICE = new PropertyDescriptor.Builder()
+            .name("Client Service")
```
Comment (jvwing) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/s3/AbstractS3Processor.java]: Name should be a machine-readable key like "client-service" with a human-readable displayName like "Client Service".  I know this has not been done previously in this file, but we should try to do it going forward.

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/PutS3Object.java [hunk 2]
```diff
@@ -154,6 +155,14 @@
         .defaultValue(StorageClass.Standard.name())
         .build();
 
+
+    public static final PropertyDescriptor PUT_ENRICHMENT_SERVICE = new PropertyDescriptor.Builder()
+            .name("Put Enrichment Service")
```
Comment (jvwing) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/PutS3Object.java]: Same request for machine-readable name and human-readable displayName.

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/EncryptedS3PutEnrichmentService.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @OnEnabled
+    public void onConfigured(final ConfigurationContext context) throws InitializationException {
+        encryptionMethod = context.getProperty(ENCRYPTION_METHOD).getValue();
+        algorithm = context.getProperty(ALGORITHM).getValue();
+        kmsKeyId = context.getProperty(KMS_KEY_ID).getValue();
```
Comment (jvwing) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/EncryptedS3PutEnrichmentService.java]: Should this evaluate expression language like `context.getProperty(KMS_KEY_ID).evaluateAttributeExpressions().getValue()`?

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/EncryptedS3ClientService.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @OnEnabled
+    public void onConfigured(final ConfigurationContext context) throws InitializationException {
+        encryptionMethod = context.getProperty(ENCRYPTION_METHOD).getValue();
+        kmsCmkId = context.getProperty(KMS_CMK_ID).getValue();
```
Comment (jvwing) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/encryption/EncryptedS3ClientService.java]: Should this evaluate expression language like `context.getProperty(KMS_CMK_ID).evaluateAttributeExpressions().getValue()`?

**Comment by jvwing** [comment]
@baank, thanks for the latest update.  Good news, we're getting down to the nit-picks:

1.  I had a checkstyle error running the full build with contrib check on nifi-aws-service-api `UnusedImports: Unused import - com.amazonaws.services.s3.AmazonS3Encryption`.

2.  In your services, some of the Property Descriptors are marked as supporting expression language, but EL is not evaluated when extracting the value of the property (like `context.getProperty(KMS_CMK_ID).evaluateAttributeExpressions().getValue()` or similar).  We should either evaluate the expressions or not mark them as supporting expression language:
  * EncryptedS3ClientService (KMS_CMK_ID, SECRET_KEY, PRIVATE_KEY, PUBLIC_KEY)
  * EncryptedS3PutEnrichmentService (KMS_KEY_ID, CUSTOMER_KEY)

It's fine to update this PR.  I'll work out rebasing and squashing when we're ready.

**Comment by baank** [comment]
@jvwing .. All changes made.

**Comment by jvwing** [comment]
@baank - thanks for the update, I think we're almost done.  Two things: 

1.) I still see a checkstyle warning for nifi-aws-processors:

> [INFO] --- maven-checkstyle-plugin:2.15:check (check-style) @ nifi-aws-processors ---
[WARNING] src/test/java/org/apache/nifi/processors/aws/s3/TestPutS3Object.java[27:8] (imports) UnusedImports: Unused import - com.amazonaws.services.s3.model.CryptoConfiguration.

Are you able to run the full build with `mvn clean install -Pcontrib-check` or otherwise run the checkstyle:check task?

2.) We still have property descriptors in EncryptedS3PutEnrichmentService that specify expression language, but are not evaluating EL (KMS_KEY_ID, CUSTOMER_KEY).  Did you mean to change those also?

**Comment by jvwing** [comment]
@baank - closed?

**Comment by jvwing** [comment]
@baank, did you mean to close this PR?  The latest changes appear to be well worth reviewing, if that is OK with you?

**Comment by baank** [comment]
We will be forking the project so please don't merge the commits.

**Pull request closed without merging**